pkg/controller/replica.go

```diff
@@ -75,104 +75,117 @@ func (r *ReconcileReplica) Reconcile(ctx context.Context, request reconcile.Requ
+				if retryCount < r.config.ControllerConfig.ReplicaVolumeAttachRetryLimit {
+					updateFunc = func(obj client.Object) error {
+						azva := obj.(*azdiskv1beta2.AzVolumeAttachment)
+						azureutils.RemoveFromMap(azva.Status.Annotations, consts.ReplicaVolumeAttachRetryAnnotation)
+						azureutils.AddToMap(azva.Status.Annotations, consts.ReplicaVolumeAttachRetryCount, strconv.Itoa(retryCount+1))
+						_, derr := updateState(azva, azdiskv1beta2.AttachmentPending, forceUpdate)
+						azva.Status.Detail = nil
+						azva.Status.Error = nil
+						return derr
+					}
+				} else {
+					updateFunc = func(obj client.Object) error {
+						azva := obj.(*azdiskv1beta2.AzVolumeAttachment)
+						azureutils.AddToMap(azva.Status.Annotations, consts.VolumeDetachRequestAnnotation, "replica-controller")
+						_, derr := updateState(azva, azdiskv1beta2.AttachmentPending, forceUpdate)
+						return derr
 					}
 				}
+
+				if _, err := azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRIStatus); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
 			}
 		}
 	}
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileReplica) handleReplicaDelete(ctx context.Context, azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment) {
-	// wait for replica AzVolumeAttachment deletion or failure of detahcment
-	waiter := r.conditionWatcher.NewConditionWaiter(ctx, watcher.AzVolumeAttachmentType, azVolumeAttachment.Name, verifyObjectFailedOrDeleted)
+func (r *ReconcileReplica) handlePromoteUpdate(ctx context.Context, azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment) {
+	// wait for the promotion update to be done and call replica management operation
+	waiter := r.conditionWatcher.NewConditionWaiter(ctx, watcher.AzVolumeAttachmentType, azVolumeAttachment.Name, verifyObjectPromotedOrDemoted)
 	defer waiter.Close()
-	_, err := waiter.Wait(ctx)
 
-	// add replica management operation to the queue if the replica AzVolumeAttachment is deleted
-	if err == nil {
+	if _, err := waiter.Wait(ctx); err == nil {
 		r.triggerManageReplica(azVolumeAttachment.Spec.VolumeName)
 	}
 }
 
+func (r *ReconcileReplica) handleDemoteUpdate(ctx context.Context, azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment) {
+	// wait for the demotion update to be done and collect garbage
+	waiter := r.conditionWatcher.NewConditionWaiter(ctx, watcher.AzVolumeAttachmentType, azVolumeAttachment.Name, verifyObjectPromotedOrDemoted)
+	defer waiter.Close()
+
+	if _, err := waiter.Wait(ctx); err == nil {
+		//nolint:contextcheck // Garbage collection is asynchronous; context is not inherited by design
+		r.triggerGarbageCollection(ctx, azVolumeAttachment.Spec.VolumeName)
+	}
+}
+
+func (r *ReconcileReplica) handleReplicaDelete(ctx context.Context, azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment, caller operationRequester) {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
```

Comment:
`GetWorkflowFromContext` will return an empty workflow as the context being fed to `handleReplicaDelete` is `context.Background()`. For such, consider using `workflow.GetWorkflow(ctx, azVolumeAttachment)` as it constructs workflow from an object if workflow cannot be found within the context.

Reply:
Or can we feed `handleReplicaDelete` with `ctx`, the parameter of reconcile function, instead of `context.Background()`?

Reply:
Using the reconciler context for a go routine would be unsafe as this context would be managed by the parent function.

Reply:
Understood, thanks for confirming.